pydicom/jsonrep.py

```diff
@@ -4,73 +4,105 @@
 import base64
 from inspect import signature
 import inspect
-from typing import Callable, Optional, Union, Any, cast
+from typing import (
+    Callable, Optional, Union, Any, cast, Type, TypeVar, Dict, TYPE_CHECKING
+)
 import warnings
 
-from pydicom.tag import BaseTag
+from pydicom.tag import BaseTag, TagType
+
+if TYPE_CHECKING:  # pragma: no cover
+    from pydicom.dataset import Dataset
 
 # Order of keys is significant!
 JSON_VALUE_KEYS = ('Value', 'BulkDataURI', 'InlineBinary',)
```

Comment:
@mrbean-bremen why [prefer](https://github.com/pydicom/pydicom/blob/506ecea8f378dc687d5c504788fc78810a190b7a/pydicom/dataset.py#L2437) BulkDataURI over InlineBinary when there are multiple keys present? Shouldn't it be the other way around?

Reply:
This comes from @pieper, I only refactored it - given that comment, I didn' touch the order (and didn't think about it). Maybe he can comment on this...

Reply:
Looks like some of the history got lost in git, but this code came from a PR with @hackermd.  I did the to_json part and he did the from_json part and the code about `JSON_VALUE_KEYS` this code is from his part.

Reply:
The implementation changed quite a bit since our PR. Not sure whether the order of the keys is indeed significant (see [Part 18 F2.2](http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_F.2.2.html)).

In DICOMweb, "metadata" is defined as data elements with a length greater than a given threshold. If the threshold is exceeded, the value is not included in the JSON object using key `Value` or `InlineBinary`, but a reference to the value is provided using key `BulkDataURI`. As far as I understand it, there should either be an `InlineBinary` or a `BulkDataURI` key, but not both, so the order shouldn't matter for any of the `from_json()` methods.

Reply:
This is what I also understand, though given that comment I wasn't sure. So I guess we can just leave it as is (and maybe remove that comment to avoid confusion).